File: packages/sdk-langgraph/agentspec_langgraph/usage_ledger.py
"""
UsageLedger — in-process token usage counter (Python mirror of TypeScript UsageLedger).

Framework-agnostic, zero-dependency accumulator for LLM token counts.
The agent code calls record() after each LLM call; the reporter drains
the ledger via snapshot(reset=True) on each heartbeat push.

Thread-safe: uses threading.Lock to protect concurrent record() calls
(common in LangGraph agents using ThreadPoolExecutor for tool calls).

Usage:
    ledger = UsageLedger()
    ledger.record("openai/gpt-4o", prompt_tokens=100, completion_tokens=50)
    snapshot = ledger.snapshot(reset=True)  # returns dict, resets counters
"""

from __future__ import annotations

import threading
from datetime import datetime, timezone
from typing import Any


# ── Private helpers ───────────────────────────────────────────────────────────


def _create_empty_entry(model_id: str) -> dict[str, Any]:
    return {
        "modelId": model_id,
        "promptTokens": 0,
        "completionTokens": 0,
        "totalTokens": 0,
        "callCount": 0,
    }


def _build_snapshot(
    counters: dict[str, dict[str, Any]],
    window_start: datetime,
) -> dict[str, Any]:
    # Shallow-copy each entry to avoid exposing mutable internal state
    models = [dict(entry) for entry in counters.values()]
    total_tokens = sum(m["totalTokens"] for m in models)
    total_calls = sum(m["callCount"] for m in models)
    return {
        "windowStartedAt": window_start.isoformat(),
        "models": models,
        "totalTokens": total_tokens,
        "totalCalls": total_calls,
    }


# ── Public class ──────────────────────────────────────────────────────────────


class UsageLedger:
    """In-process token usage accumulator.

    Thread-safe. O(1) record(), snapshot returns a dict matching the
    TypeScript UsageSnapshot shape.
    """

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._counters: dict[str, dict[str, Any]] = {}
        self._window_start = datetime.now(timezone.utc)

    def record(
        self,
        model_id: str,
        prompt_tokens: int = 0,
        completion_tokens: int = 0,
    ) -> None:
        """Record token usage for a single LLM call."""
        p = max(0, prompt_tokens)
        c = max(0, completion_tokens)
        with self._lock:
            entry = self._counters.get(model_id)
            if entry is None:
                entry = _create_empty_entry(model_id)
                self._counters[model_id] = entry
            entry["promptTokens"] += p
            entry["completionTokens"] += c
            entry["totalTokens"] += p + c
            entry["callCount"] += 1

    def snapshot(self, reset: bool = True) -> dict[str, Any]:
        """Return a frozen snapshot of the current usage window.

        Args:
            reset: When True (default), clears all counters and starts a new window.
        """
        with self._lock:
            snap = _build_snapshot(self._counters, self._window_start)
            if reset:
                self._counters = {}
                self._window_start = datetime.now(timezone.utc)
        return snap

File: packages/sdk/src/agent/usage-ledger.ts
/** In-process token usage counter — aggregates LLM call tokens between heartbeat pushes. */

// ── Types (exported) ─────────────────────────────────────────────────────────

export interface ModelUsageEntry {
  modelId: string
  promptTokens: number
  completionTokens: number
  totalTokens: number
  callCount: number
}

export interface UsageSnapshot {
  windowStartedAt: string
  models: ModelUsageEntry[]
  totalTokens: number
  totalCalls: number
}

// ── Private helpers ──────────────────────────────────────────────────────────

function createEmptyEntry(modelId: string): ModelUsageEntry {
  return { modelId, promptTokens: 0, completionTokens: 0, totalTokens: 0, callCount: 0 }
}

function buildSnapshot(
  counters: Map<string, ModelUsageEntry>,
  windowStart: Date,
): UsageSnapshot {
  // Shallow-copy each entry to avoid exposing mutable internal state
  const models = [...counters.values()].map((e) => ({ ...e }))
  let totalTokens = 0
  let totalCalls = 0
  for (const m of models) {
    totalTokens += m.totalTokens
    totalCalls += m.callCount
  }
  return {
    windowStartedAt: windowStart.toISOString(),
    models,
    totalTokens,
    totalCalls,
  }
}

// ── Public class ─────────────────────────────────────────────────────────────

export class UsageLedger {
  private counters = new Map<string, ModelUsageEntry>()
  private windowStart = new Date()

  /**
   * Record token usage for a single LLM call.
   * Call this once per LLM invocation with the model ID and token counts.
   */
  record(modelId: string, promptTokens: number, completionTokens: number): void {
    const p = Math.max(0, promptTokens)
    const c = Math.max(0, completionTokens)
    let entry = this.counters.get(modelId)
    if (!entry) {
      entry = createEmptyEntry(modelId)
      this.counters.set(modelId, entry)
    }
    entry.promptTokens += p
    entry.completionTokens += c
    entry.totalTokens += p + c
    entry.callCount += 1
  }

  /**
   * Return a frozen snapshot of the current usage window.
   *
   * @param reset - When true (default), clears all counters and starts a new
   *   window. Use reset=true in the heartbeat push so each window is shipped
   *   exactly once. Use reset=false for inspection without draining.
   */
  snapshot(reset = true): UsageSnapshot {
    const snap = buildSnapshot(this.counters, this.windowStart)
    if (reset) {
      this.counters = new Map()
      this.windowStart = new Date()
    }
    return snap
  }
}
